src/app.js

```diff
@@ -7,6 +7,19 @@ export function app(app) {
   var node
   var element
 
+  function hydrate(elm) {
+    return elm == null ? undefined : {
+      tag: elm.tagName,
+      data: {},
+      children: [].map.call(elm.childNodes, function(child){hydrate(child)})
+    }
+  }
+
+  if (root.hasChildNodes !== undefined && root.hasChildNodes()) {
+    element = root.children[0]
+    node = hydrate(element)
+  }
+
```

Comment:
That is why I have been asking for clarification on the non-empty root test case.  Is that the expected behavior, if root is not empty then append to children?  That is not what I would expect, and it makes hydration impossible just by contradicting the needed behavior I believe.  I am not aware of react/preact/inferno/Vue having this append to children if root isn't empty behavior, but they do hydrate existing dom from ssr to prevent flicker from removing existing dom before rendering.  

I think it will be clearer if I show a video of the behavior of hyperapp before and after this patch.  I am making one now to share for clarification.